documentation/design-docs/diagnostics-client-library.md

```diff
@@ -0,0 +1,329 @@
+                }
+            }
+        }
+    }
+    try
+    {
+        source.Process();
+        shouldExit.WaitOne();
+    }
+    // TraceEvent throws a generic Exception when the target process exists first. 
+    // This also needs some fix on TraceEvent side.
+    catch (Exception e) { } 
+    finally
+    {
+        DiagnosticsClient.StopTracing(session);
+    }
+}
+```
+
+#### 3. Trigger a CPU trace when CPU usage goes above a certain threshold
+```cs
+
+using Microsoft.Diagnostics.NETCore.Client;
+using System.Diagnostics;
+using System.IO;
+using System.Threading.Task;
+
+public void TriggerTraceOnCpuUsage(int processId, int threshold, int traceDuration, string traceName)
+{
+    IEnumerable<EventPipeProvider> runtimeCounterProvider = new List<EventPipeProvider>() { 
+        new EventPipeProvider("System.Runtime", 1, 1, "EventCounterIntervalSec=1")
+    };
+
+    IEnumerable<EventPipeProvider> cpuProvider = new List<EventPipeProvider>() { 
+        new EventPipeProvider("Microsoft-Windows-DotNETRuntime", ClrTraceEventParser.Keywords.Default, EventLevel.Informational)
+    };
+
+    EventPipeSession counterSession = DiagnosticsClient.StartTracing(processId, runtimeCounterProvider);
+    EventPipeEventSource source = new EventPipeEventSource(counterSession.stream)
+    source.Dynamic.All += (TraceEvent obj) => {
+        if (obj.EventName.Equals("EventCounters"))
+        {
+            // I know this part is ugly. But this is all TraceEvent.
+            IDictionary<string, object> payloadFields = (IDictionary<string, object>)(obj.GetPayloadValueByName("Payload"));
+            if (payloadFields["Name"].ToString().Equals("cpu-usage"))
+            {
+                double cpuUsage = Double.Parse(payloadFields["Mean"]);
+                if (cpuUsage > (double)threshold)
+                {
+                    EventPipeSession traceSession = DiagnosticsClient.StartTracing(processId, cpuProvider);
+                    if (traceSesssion != null)
+                    {
+                        await Task.Run(() => {
+                            var buffer = new byte[16 * 1024];
+                            var fs = new FileStream(traceName, FileMode.Create, FileAccess.Write))
+                            Stopwatch sw = new Stopwatch();
+                            sw.Start();
+                            while (sw.Elapsed.Seconds < traceDuration)
+                            {
+                                int nBytesRead = traceSession.Stream.Read(buffer, 0, buffer.Length);
+                                if (nBytesRead <= 0)
+                                    break;
+                                fs.Write(buffer, 0, nBytesRead);
+                            }
+                            DiagnosticsClient.StopTracing(traceSession);
+                            shouldExit.Set();
+                        });
+                    }
+                }
+            }
+        }
+    }
+    try
+    {
+        source.Process();
```

Comment:
I don't think you stop the `counterSession`, so I don't think this call will ever return.